### PR TITLE
Add JsonRpcProxy

### PR DESCRIPTION
**Pull request by oo7ww**
A Json Rpc Proxy 
* support websocket and http url
* use json-rpc-engine as middleware

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -11,9 +11,12 @@
   "author": "",
   "license": "ISC",
   "dependencies": {
+    "axios": "^0.21.1",
     "debug": "^4.1.1",
     "js-conflux-sdk": "^1.5.11",
+    "json-rpc-engine": "^6.1.0",
     "lodash": "^4.17.21",
+    "pify": "^5.0.0",
```
Comment (Pana) [inline package.json:19]: Seems axios and pify is not used in this project right ?

File: src/JsonRpcProxy/Adaptor.js [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+const defaultMethodAdaptor = require('../defaultMethodAdaptor');
+const format = require('../format');
+const _ = require('lodash');
+const util = require('../util');
+const { Conflux, Transaction } = require('js-conflux-sdk');
+
+
+class Proxy {
+    constructor(url, networkId) {
+        this.cfx = new Conflux({url: url});
```
Comment (Pana) [inline src/JsonRpcProxy/Adaptor.js:11]: networkId should been used to init Conflux object

File: src/JsonRpcProxy/Adaptor.js [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+
+    async function estimateGas(req, res, next, end) {
+        req.method = defaultMethodAdaptor(req.method, req.params);
+        req.params = format.formatCommonInput(req.params, proxy.networkId);
+        await next();
+        if(!res || !res.result) return;
+        res.result = res.result.gasUsed;
+    }
+
+    function methodAdaptor(method, params) {
```
Comment (Pana) [inline src/JsonRpcProxy/Adaptor.js:259]: This method's name is a little confusing with above adaptMethod

File: src/JsonRpcProxy/send.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
+const { createAsyncMiddleware } = require('json-rpc-engine');
+const { providerFactory } = require('js-conflux-sdk');
+
+class Agent {
```
Comment (Pana) [inline src/JsonRpcProxy/send.js:4]: Seems we can omit this Class here, direct create an provider below

File: src/JsonRpcProxy/Adaptor.js [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+            "epochNumber",
+            "index",
+            "outcomeStatus",
+            "stateRoot"
+        ]);
+        return res;
+    }
+
+    async function sendTransaction(req, res, next) {
+        req.method = methodAdaptor(req.method, req.params);
```
Comment (Pana) [inline src/JsonRpcProxy/Adaptor.js:225]: Some fields of an transaction parameter can be filled automatically for example: nonce, gas, gasPrice, chainId and so on

**Review by Pana — COMMENTED** [review]
There are couple issues should be fixed

File: package.json (hunk 1 above)
Comment (oo7ww) [reply package.json]: rm unused modules

File: src/JsonRpcProxy/Adaptor.js (hunk 2 above)
Comment (oo7ww) [reply src/JsonRpcProxy/Adaptor.js]: networkId added

File: src/JsonRpcProxy/Adaptor.js (hunk 3 above)
Comment (oo7ww) [reply src/JsonRpcProxy/Adaptor.js]: renamed

File: src/JsonRpcProxy/send.js (hunk 4 above)
Comment (oo7ww) [reply src/JsonRpcProxy/send.js]: modified

File: src/JsonRpcProxy/Adaptor.js (hunk 5 above)
Comment (oo7ww) [reply src/JsonRpcProxy/Adaptor.js]: fixed

**Pull request merged**